runtime-transaction/src/signature_details.rs

```diff
@@ -10,42 +10,64 @@ pub struct PrecompileSignatureDetails {
-    let mut num_secp256k1_instruction_signatures: u64 = 0;
-    let mut num_ed25519_instruction_signatures: u64 = 0;
-    let mut num_secp256r1_instruction_signatures: u64 = 0;
-    for (program_id, instruction) in instructions {
+pub(crate) struct PrecompileSignatureDetailsBuilder {
+    filter: SignatureDetailsFilter,
+    value: PrecompileSignatureDetails,
+}
+
+impl Default for PrecompileSignatureDetailsBuilder {
+    fn default() -> Self {
+        Self {
+            filter: SignatureDetailsFilter::new(),
+            value: PrecompileSignatureDetails {
+                num_secp256k1_instruction_signatures: 0,
+                num_ed25519_instruction_signatures: 0,
+                num_secp256r1_instruction_signatures: 0,
+            },
+        }
+    }
+}
+
+impl PrecompileSignatureDetailsBuilder {
+    pub fn process_instruction(&mut self, program_id: &Pubkey, instruction: &SVMInstruction) {
         let program_id_index = instruction.program_id_index;
-        match filter.is_signature(program_id_index, program_id) {
+        match self.filter.is_signature(program_id_index, program_id) {
             ProgramIdStatus::NotSignature => {}
             ProgramIdStatus::Secp256k1 => {
-                num_secp256k1_instruction_signatures = num_secp256k1_instruction_signatures
-                    .wrapping_add(get_num_signatures_in_instruction(&instruction));
+                self.value.num_secp256k1_instruction_signatures = self
+                    .value
+                    .num_secp256k1_instruction_signatures
+                    .wrapping_add(get_num_signatures_in_instruction(instruction));
             }
             ProgramIdStatus::Ed25519 => {
-                num_ed25519_instruction_signatures = num_ed25519_instruction_signatures
-                    .wrapping_add(get_num_signatures_in_instruction(&instruction));
+                self.value.num_ed25519_instruction_signatures = self
+                    .value
+                    .num_ed25519_instruction_signatures
+                    .wrapping_add(get_num_signatures_in_instruction(instruction));
             }
             ProgramIdStatus::Secp256r1 => {
-                num_secp256r1_instruction_signatures = num_secp256r1_instruction_signatures
-                    .wrapping_add(get_num_signatures_in_instruction(&instruction));
+                self.value.num_secp256r1_instruction_signatures = self
+                    .value
+                    .num_secp256r1_instruction_signatures
+                    .wrapping_add(get_num_signatures_in_instruction(instruction));
             }
         }
     }
 
-    PrecompileSignatureDetails {
-        num_secp256k1_instruction_signatures,
-        num_ed25519_instruction_signatures,
-        num_secp256r1_instruction_signatures,
+    pub fn build(self) -> PrecompileSignatureDetails {
+        self.value
+    }
+}
+
+/// Get transaction signature details.
+pub fn get_precompile_signature_details<'a>(
```

Comment:
left this as public since there's a benchmark for this function. may be useful going forward to have these separated so we can bench individually. but may also just remove this and have a single benchmark for instruction meta.